Reuse bot object in API and optionally use chronology protector

micgro42 opened the pull request:
Using the chronology protector should prevent some browser test failures when running against beta and the replica hasn't caught up yet.

TODO: Somehow try this out in my local browser tests

Bug: [T277862](https://phabricator.wikimedia.org/T277862)
Bug: [T284443](https://phabricator.wikimedia.org/T284443)

micgro42 committed:
Reuse bot object in API and optionally use chronology protector

Using the chronology protector should prevent some browser test failures
when running against beta and the replica hasn't caught up yet.

Bug: T277862
Bug: T284443

micgro42 commented:
Oh, apparently that is a repository where we still do have some Travis-CI integration? o.O

lucaswerkmeister commented on wikibase.api.js:
```diff
@@ -28,51 +71,39 @@ class WikibaseApi {
-			} ).then( ( response ) => {
-				return response.entity.id;
-			} );
+		return bot.request( {
```
Won’t this fail because `bot` is a Promise instead of an MWBot? We’d need something like this, I think:

```js
return this.getBot()
    .then( ( bot ) => bot.request( {
        action: 'wbeditentity',
```

micgro42 replied:
True. OTOH, I wonder if we can use `async`/`await` by now and just have `const bot = await this.getBot();`?

lucaswerkmeister replied:
Yeah, it looks like async/await is available in Node 12 (or even Node 10 already). Might need a change to some eslint configs, but apart from that it should be fine, I think.

lucaswerkmeister reviewed (commented):
> TODO: Somehow try this out in my local browser tests

https://stackoverflow.com/a/39732501/1420237 ?

micgro42 commented:
> > TODO: Somehow try this out in my local browser tests
> 
> [stackoverflow.com/a/39732501/1420237](https://stackoverflow.com/a/39732501/1420237) ?

I tried it locally with [npm-link](https://docs.npmjs.com/cli/v7/commands/npm-link), will make a patch for Gerrit/Jenkins with this approach!

Edit: https://gerrit.wikimedia.org/r/c/mediawiki/extensions/WikibaseLexeme/+/738378 though I may have to chain this behind https://gerrit.wikimedia.org/r/c/mediawiki/extensions/WikibaseLexeme/+/709747 to make all the peer dependencies line up correctly

lucaswerkmeister commented on wikibase.api.js:
```diff
@@ -28,51 +71,39 @@ class WikibaseApi {
-					}, reject );
-				} );
-			} );
+		return new Promise( ( resolve, reject ) => { // FIXME: I don't think this Promise is needed
```
I don’t think the Promise is needed either, we can probably just await the response and then `return response.entity.id`? (And I’d say the same for `createItem()` – once we start using async/await, it seems unnecessary to keep working with `.then()` “manually”.)

micgro42 replied:
Yes. Though I feel like I want to split up the change to how we use mwbot and general Promise=>async/await cleanup into two commits

lucaswerkmeister commented on wikibase.api.js:
```diff
@@ -97,10 +126,6 @@ class WikibaseApi {
-				username: browser.config.mwUser,
-				password: browser.config.mwPwd
-			} );
 		} ).then( () => {
```
This line can probably be removed as well?

micgro42 replied:
yes, done!

micgro42 committed:
Simplify Api code by replacing Promises with async/await

This should be a pure refactoring without any behavior changes

micgro42 committed:
Minor version bump

This is not a major version bump, because everything still works without
using `WikibaseApi.initialize()`, it just logs a warning.

lucaswerkmeister reviewed (approved):
Wow, even Node v8 had async/await already, we don’t even need to bump the engine version in `package.json`.

LGTM; do we need more testing (CI somewhere), or is this good to release?

micgro42 commented:
> LGTM; do we need more testing (CI somewhere), or is this good to release?

I _think_ this is good to go, but to be sure I think I would like the Wikibase patch to be ready similar to the Lexeme patch before merging it.

We should make a ticket for replacing Travis in this repo here

lucaswerkmeister commented:
Ah, okay. So we’re waiting for CI to finish on https://gerrit.wikimedia.org/r/c/mediawiki/extensions/Wikibase/+/738421, I assume. (And, yes!)

micgro42 commented:
Both the Wikibase and WikibaseLexeme patches succeeded, so this is ready to be merged and a new release made
